### PR TITLE
sinkv2(ticdc): activite sinkv2 mysql sink

### PR DESCRIPTION
**Pull request by hicqu**
### What problem does this PR solve?

Issue Number: ref #5928 

### What is changed and how it works?

Use sinkv2 mysql sink.

#### Tests 

 - Unit test

### Release note 

```release-note
None
```

### CODE REVIEW

File: cdc/sink/mysql/mysql.go [hunk 1]
```diff
@@ -818,7 +818,7 @@ func (s *mysqlSink) prepareDMLs(rows []*model.RowChangedEvent, bucket int) *prep
 func (s *mysqlSink) execDMLs(ctx context.Context, rows []*model.RowChangedEvent, bucket int) error {
 	failpoint.Inject("SinkFlushDMLPanic", func() {
```
Comment (hi-rustin) [inline cdc/sink/mysql/mysql.go:819]: Please remove this failpoint.

File: cdc/sinkv2/eventsink/event_appender.go [hunk 2]
```diff
@@ -69,26 +69,16 @@ func (t *TxnEventAppender) Append(
...
-					zap.Any("row", row))
-			}
-			startTsDecreased := lastTxn.StartTs > row.StartTs
-			if startTsDecreased {
-				log.Panic("The startTs of the emit row is less than the received row",
-					zap.Any("lastReceivedStartTs", buffer[len(buffer)-1].GetCommitTs()),
-					zap.Any("row", row))
-			}
 
+		lastCommitTs := lastTxn.GetCommitTs()
```
Comment (hi-rustin) [inline cdc/sinkv2/eventsink/event_appender.go:73]: Why do we remove other checks?

File: cdc/sinkv2/eventsink/txn/mysql/mysql.go [hunk 3]
```diff
@@ -114,9 +115,13 @@ func (s *mysqlBackend) OnTxnEvent(event *eventsink.TxnCallbackableEvent) (needFl
 
 // Flush implements interface backend.
 func (s *mysqlBackend) Flush(ctx context.Context) (err error) {
+	if s.rows == 0 {
+		return
+	}
+
 	failpoint.Inject("SinkFlushDMLPanic", func() {
```
Comment (hi-rustin) [inline cdc/sinkv2/eventsink/txn/mysql/mysql.go:122]: Please remove this failpoint.

File: tests/integration_tests/processor_resolved_ts_fallback/run.sh [hunk 4]
```diff
@@ -1,63 +1 @@
-#!/bin/bash
```
Comment (hi-rustin) [inline tests/integration_tests/processor_resolved_ts_fallback/run.sh:1]: Please revert this change.

**Review by hi-rustin — COMMENTED** [review]
rest LGTM

File: cdc/sinkv2/eventsink/event_appender.go (hunk 2 above)
Comment (hicqu) [reply cdc/sinkv2/eventsink/event_appender.go:73]: startTs can fallback, it's expected. So I removed it.

**Comment by hi-rustin** [comment]
@hicqu The github actions failed.

File: cdc/sinkv2/eventsink/blackhole/black_hole_dml_sink.go [hunk 5]
```diff
@@ -33,7 +33,8 @@ func New() *sink {
 // WriteEvents log the events.
 func (s *sink) WriteEvents(rows ...*eventsink.CallbackableEvent[*model.RowChangedEvent]) error {
 	for _, row := range rows {
-		log.Debug("BlockHoleSink: WriteEvents", zap.Any("row", row))
+		// NOTE: don't change the log, some tests depend on it.
+		log.Debug("BlockHoleSink: EmitRowChangedEvents", zap.Any("row", row.Event))
```
Comment (hi-rustin) [inline cdc/sinkv2/eventsink/blackhole/black_hole_dml_sink.go:37]: I think we should change the log checker to check the new name.

File: cdc/sinkv2/eventsink/event_appender.go [hunk 6]
```diff
@@ -69,26 +69,21 @@ func (t *TxnEventAppender) Append(
...
+				zap.Any("row", row))
+		} else if lastCommitTs == row.CommitTs && lastTxn.StartTs > row.StartTs {
+			log.Panic("The startTs of the emit row is less than the received row with same CommitTs",
+				zap.Uint64("lastReceivedCommitTs", lastCommitTs),
+				zap.Uint64("lastReceivedStartTs", lastTxn.StartTs),
+				zap.Any("row", row))
+		}
+
+		// Split on big transactions or a new one.
+		if row.SplitTxn || lastTxn.StartTs != row.StartTs {
```
Comment (hi-rustin) [inline cdc/sinkv2/eventsink/event_appender.go:86]: Should we append a new txn when the committs grow?

File: cdc/sinkv2/eventsink/factory/factory.go [hunk 7]
```diff
@@ -53,8 +55,18 @@ func New(ctx context.Context,
 
 	s := &SinkFactory{}
 	schema := strings.ToLower(sinkURI.Scheme)
-	// TODO: add more sink factory here.
 	switch schema {
+	case sink.MySQLSchema, sink.MySQLSSLSchema, sink.TiDBSchema, sink.TiDBSSLSchema:
+		sinkURI, err := url.Parse(sinkURIStr)
```
Comment (hi-rustin) [inline cdc/sinkv2/eventsink/factory/factory.go:60]: SinkURI already at the top?

File: cdc/sinkv2/eventsink/txn/mysql/mysql.go [hunk 8]
```diff
@@ -88,21 +87,23 @@ func NewMySQLBackend(
...
+			statistics:  statistics,
+		})
 	}
 
-	log.Info("mysql backend is created",
-		zap.String("changefeedID", fmt.Sprintf("%s.%s", changefeedID.Namespace, changefeedID.ID)),
-		zap.Bool("forceReplicate", s.cfg.ForceReplicate),
-		zap.Bool("enableOldValue", s.cfg.EnableOldValue))
-	return s, nil
+	log.Info("mysql backends is created",
```
Comment (hi-rustin) [inline cdc/sinkv2/eventsink/txn/mysql/mysql.go:101]:
```suggestion
	log.Info("MySQL backends is created",
```

File: cdc/sinkv2/eventsink/txn/mysql/mysql.go [hunk 9]
```diff
@@ -367,28 +361,21 @@ func (s *mysqlBackend) execDMLWithMaxRetries(ctx context.Context, dmls *prepared
...
 	query string, count int, startTs []model.Ts,
 ) error {
 	if isRetryableDMLError(err) {
 		log.Warn("execute DMLs with error, retry later",
 			zap.Error(err), zap.Duration("duration", time.Since(start)),
 			zap.String("query", query), zap.Int("count", count),
 			zap.Uint64s("startTs", startTs),
-			zap.String("namespace", changefeed.Namespace),
-			zap.String("changefeed", changefeed.ID))
+			zap.String("changefeed", changefeed))
```
Comment (hi-rustin) [inline cdc/sinkv2/eventsink/txn/mysql/mysql.go:373]: This is different from our other logs.

File: cdc/sinkv2/eventsink/txn/txn_sink_test.go [hunk 10]
```diff
@@ -245,3 +257,63 @@ func TestGenKeys(t *testing.T) {
...
+	}
+
+	// Wait a while so that the first worker can meet the error.
+	errCh := make(chan error, 1)
+	sink := newSink(ctx, bes, errCh, DefaultConflictDetectorSlots)
+	time.Sleep(100 * time.Millisecond)
+	select {
+	case <-errCh:
+	default:
+		log.Fatal("a flush error is expected")
```
Comment (hi-rustin) [inline cdc/sinkv2/eventsink/txn/txn_sink_test.go:286]: Could you please use t.Fatal in the tests?

File: cdc/sinkv2/eventsink/txn/txn_sink_test.go [hunk 11]
```diff
@@ -245,3 +257,63 @@ func TestGenKeys(t *testing.T) {
...
+	bes := make([]backend, 0, 1)
+	bes = append(bes, &blackhole{panicOnFlush: 1})
+	errCh := make(chan error, 1)
+	sink := newSink(context.Background(), bes, errCh, DefaultConflictDetectorSlots)
+
+	time.Sleep(1 * time.Second)
+	select {
+	case <-errCh:
+	default:
+		panic("should get an error")
```
Comment (hi-rustin) [inline cdc/sinkv2/eventsink/txn/txn_sink_test.go:316]: Can we use t.Fatal or t.Fail?

File: cdc/sinkv2/eventsink/txn/txn_sink_test.go [hunk 12]
```diff
@@ -16,26 +16,30 @@ package txn
...
+	"github.com/pingcap/log"
 	"github.com/pingcap/tidb/parser/mysql"
 	"github.com/pingcap/tiflow/cdc/model"
 	"github.com/pingcap/tiflow/cdc/sinkv2/eventsink"
-	"github.com/pingcap/tiflow/pkg/notify"
+	mock_txn "github.com/pingcap/tiflow/cdc/sinkv2/eventsink/txn/mock"
 	"github.com/stretchr/testify/require"
 )
 
 type blackhole struct {
```
Comment (hi-rustin) [inline cdc/sinkv2/eventsink/txn/txn_sink_test.go:34]: Maybe we can rename it mockBackend because we have a blackhole sink in sinkv2. Perhaps it will help to reduce the confusion.

File: cdc/sinkv2/eventsink/txn/txn_sink_test.go [hunk 13]
```diff
@@ -245,3 +257,63 @@ func TestGenKeys(t *testing.T) {
...
+	case <-errCh:
+	default:
+		log.Fatal("a flush error is expected")
+	}
+
+	// sink.Close can close all background goroutines. No goroutines should be leak.
+	require.Nil(t, sink.Close())
+	select {
+	case <-errCh:
+		log.Fatal("no error is expected")
```
Comment (hi-rustin) [inline cdc/sinkv2/eventsink/txn/txn_sink_test.go:293]: ditto

File: cdc/sinkv2/eventsink/txn/worker.go [hunk 14]
```diff
@@ -38,15 +45,23 @@ type worker struct {
...
+	timer         *time.Timer
+	flushInterval time.Duration
 }
 
-func newWorker(ID int, backend backend, errCh chan<- error) *worker {
+func newWorker(ctx context.Context, ID int, backend backend, errCh chan<- error) *worker {
+	changefeedID := contextutil.ChangefeedIDFromCtx(ctx)
 	return &worker{
+		ctx:        ctx,
+		changefeed: fmt.Sprintf("%s.%s", changefeedID.Namespace, changefeedID.ID),
```
Comment (hi-rustin) [inline cdc/sinkv2/eventsink/txn/worker.go:56]: This is different from our other logs.

File: cdc/sinkv2/eventsink/txn/worker.go [hunk 15]
```diff
@@ -67,49 +82,91 @@ func (w *worker) runBackgroundLoop() {
...
+				log.Info("Transaction sink backend close fail",
+					zap.String("changefeedID", w.changefeed),
+					zap.Int("workerID", w.ID),
 					zap.Error(err))
 			}
 		}()
-		w.timer = time.NewTimer(w.backend.MaxFlushInterval())
+		defer func() {
+			var r interface{}
+			if r = recover(); r == nil {
```
Comment (hi-rustin) [inline cdc/sinkv2/eventsink/txn/worker.go:93]:
```suggestion
			if r = recover(); r == nil {
```
Should we print some logs here?

File: cdc/sinkv2/eventsink/txn/worker.go [hunk 16]
```diff
@@ -67,49 +82,91 @@ func (w *worker) runBackgroundLoop() {
...
 				return
 			case txn := <-w.txnCh.Out():
 				metrics.ConflictDetectDuration.Observe(time.Since(txn.start).Seconds())
-				txn.wantMore()
-				if w.backend.OnTxnEvent(txn.txnEvent.TxnCallbackableEvent) && w.doFlush() {
-					log.Warn("Transaction sink backend exits unexceptedly")
-					return
+				wantMoreCallbacks = append(wantMoreCallbacks, txn.wantMore)
+				if w.backend.OnTxnEvent(txn.txnEvent.TxnCallbackableEvent) {
+					if w.doFlush() {
```
Comment (hi-rustin) [inline cdc/sinkv2/eventsink/txn/worker.go:125]: Could you please add some comments to explain what happened and why we need to break or continue?
Also, should we add some logs here?

File: tests/integration_tests/processor_resolved_ts_fallback/run.sh [hunk 17]
```diff
@@ -18,7 +18,7 @@ function run() {
 	start_tidb_cluster --workdir $WORK_DIR
 	cd $WORK_DIR
 
-	export GO_FAILPOINTS='github.com/pingcap/tiflow/cdc/sink/mysql/SinkFlushDMLPanic=return(true);github.com/pingcap/tiflow/cdc/sink/mq/producer/kafka/SinkFlushDMLPanic=return(true)'
+	export GO_FAILPOINTS='github.com/pingcap/tiflow/cdc/sinkv2/eventsink/txn/mysql/MySQLSinkExecDMLError=return(true);github.com/pingcap/tiflow/cdc/sinkv2/eventsink/mq/dmlproducer/KafkaSinkAsyncSendError=return(true)'
```
Comment (hi-rustin) [inline tests/integration_tests/processor_resolved_ts_fallback/run.sh:21]: Can we remove this?

**Review by hi-rustin — COMMENTED** [review]
GitHub actions still failed. I think it's because we missing the `SimpleMySQLSinkTester` failpoint. Maybe we can just delete this old value test because we have some other old value tests.

rest LGTM.

File: tests/integration_tests/processor_resolved_ts_fallback/run.sh (hunk 17 above)
Comment (hicqu) [reply tests/integration_tests/processor_resolved_ts_fallback/run.sh:21]: No. We expect that the cdc instance can't execute SQLs before gets killed.

File: cdc/sinkv2/eventsink/event_appender.go (hunk 6 above)
Comment (hicqu) [reply cdc/sinkv2/eventsink/event_appender.go:86]: If commit timestamps are different, their start timestamps must also be different, too.

File: cdc/sinkv2/eventsink/txn/mysql/mysql.go (hunk 9 above)
Comment (hicqu) [reply cdc/sinkv2/eventsink/txn/mysql/mysql.go:373]: I think both are OK so I prefer to let it be.

File: cdc/sinkv2/eventsink/txn/worker.go (hunk 15 above)
Comment (hicqu) [reply cdc/sinkv2/eventsink/txn/worker.go:93]: There is already a log for existing from the for loop.

File: cdc/sinkv2/eventsink/txn/worker.go (hunk 16 above)
Comment (hicqu) [reply cdc/sinkv2/eventsink/txn/worker.go:125]: If you mean the meanning of `doFlush` return value, It has already been commentted on `doFlush` itself.

File: cdc/sinkv2/eventsink/txn/worker.go [hunk 18]
```diff
@@ -67,49 +82,91 @@ func (w *worker) runBackgroundLoop() {
...
 		case w.errCh <- err:
-		case <-ctx.Done():
 		}
 		return true
 	}
 	if !w.timer.Stop() {
-		<-w.timer.C
+		select {
+		case <-w.timer.C:
+		default:
```
Comment (hi-rustin) [inline cdc/sinkv2/eventsink/txn/worker.go:167]: Why do we need a default branch here?

File: cdc/sinkv2/eventsink/event_appender.go (hunk 6 above)
Comment (hi-rustin) [reply cdc/sinkv2/eventsink/event_appender.go:86]: Could you please add a comment here?

File: cdc/sinkv2/eventsink/txn/txn_sink_test.go (hunk 10 above)
Comment (hi-rustin) [inline cdc/sinkv2/eventsink/txn/txn_sink_test.go:286]:
```suggestion
		t.Fatal("a flush error is expected")
```

File: cdc/sinkv2/eventsink/txn/txn_sink_test.go (hunk 13 above)
Comment (hi-rustin) [inline cdc/sinkv2/eventsink/txn/txn_sink_test.go:293]:
```suggestion
		t.Fatal("no error is expected")
```

File: cdc/sinkv2/eventsink/txn/txn_sink_test.go (hunk 11 above)
Comment (hi-rustin) [inline cdc/sinkv2/eventsink/txn/txn_sink_test.go:316]:
```suggestion
		t.Fatal("should get an error")
```

File: cdc/sinkv2/eventsink/txn/mysql/mysql.go (hunk 9 above)
Comment (hi-rustin) [reply cdc/sinkv2/eventsink/txn/mysql/mysql.go:373]: It might be better to unify the log structure. No matter, we need to add string method for changefeedID to unify it. We can modify it in the future.

File: cdc/sinkv2/eventsink/txn/worker.go (hunk 18 above)
Comment (hicqu) [reply cdc/sinkv2/eventsink/txn/worker.go:167]: Yes. Other wise it will be blocked if the timer has already been triggered.

**Review by hi-rustin — APPROVED** [review]

**Review by nongfushanquan — COMMENTED** [review]
why remove some failpoint - injects in this pr

**Review by nongfushanquan — APPROVED** [review]
